### PR TITLE
[consensus] shortcut the block retrieval with local buffer

### PR DESCRIPTION
**Pull request by zekun000**
## Type of Change
- [ ] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## Checklist
- [ ] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I identified and added all stakeholders and component owners affected by this change as reviewers
- [ ] I tested both happy and unhappy path of the functionality
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: consensus/src/block_storage/pending_blocks.rs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    pub fn insert_block(&mut self, block: Block) {
+        info!("Pending block inserted: {}", block.id());
+        self.blocks_by_hash.insert(block.id(), block.clone());
+        self.blocks_by_round.insert(block.round(), block.clone());
+        if let Some((id, tx)) = self.pending_request.take() {
+            if id == block.id() {
+                info!("FulFill block request from incoming block: {}", id);
+                BLOCK_RETRIEVAL_LOCAL_FULFILL_COUNT.inc();
```
Comment (sitalkedia) [inline consensus/src/block_storage/pending_blocks.rs:35]: Are you seeing the pending blocks being fulfilled in the Forge run?

File: consensus/src/block_storage/sync_manager.rs [hunk 2]
```diff
@@ -443,6 +449,28 @@ impl BlockRetriever {
         monitor!("retrieve_block_for_id_chunk", {
             let mut interval = time::interval(retry_interval);
             let mut futures = FuturesUnordered::new();
+            if retrieve_batch_size == 1 {
```
Comment (sitalkedia) [inline consensus/src/block_storage/sync_manager.rs:452]: Why do this only for 1 batch size?

File: consensus/src/block_storage/pending_blocks.rs (hunk 1 above)
Comment (zekun000) [reply consensus/src/block_storage/pending_blocks.rs:35]: yes, in forge all of them are fulfilled by the local buffer

File: consensus/src/block_storage/sync_manager.rs (hunk 2 above)
Comment (zekun000) [reply consensus/src/block_storage/sync_manager.rs:452]: we only do multi block fetch during the fast forward sync process, which seems unlikely to hit the local buffer

File: consensus/src/block_storage/pending_blocks.rs (hunk 1 above)
Comment (sitalkedia) [reply consensus/src/block_storage/pending_blocks.rs:35]: Are you seeing any top-level metrics improvement like latency or throughput?

**Review by sitalkedia — APPROVED** [review]
So excited about this feature. Looking forward to seeing this win in production.

**Review by danielxiangzl — APPROVED** [review]

**Pull request merged**